### ISSUE TITLE
`unavailable_function` and overriding

### ISSUE THREAD

**Issue by hajduakos** [realm/SwiftLint#6222]
### New Issue Checklist

- [X] I've Updated SwiftLint to the latest version.
- [X] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Feature or Enhancement Proposal

The `unavailable_function` rule reports functions which only have a terminating call (e.g. `fatalError`) without any return, unless marked `available(*, unavailable)` or returning `Never`. However, this can cause false positives when functions are overridden. Consider the following example (`main.swift`):
```Swift
class Base {
    func greet() { fatalError("Subclasses must override greet()") }
}
class Derived: Base {
    override func greet() { print("Hello World") }
}
Derived().greet()
```
and `.swiftlint.yml`
```yaml
opt_in_rules:
  - unavailable_function
 ```
Running `swiftlint .` will report:
```
main.swift:2:5: warning: Unavailable Function Violation: Unimplemented functions should be marked as unavailable (unavailable_function)
```

However, adding `@available(*, unavailable)` to `greet` in `Base` like this
```Swift
class Base {
    @available(*, unavailable)
    func greet() { fatalError("Subclasses must override greet()") }
}
class Derived: Base {
    override func greet() { print("Hello World") }
}
Derived().greet()
```
will result in a compile error:
```
main.swift:6:19 Cannot override 'greet' which has been marked unavailable
```

I'm wondering if it would make sense to make this rule stricter and avoid reporting such cases, also considering visibility:
- `open`: I don't think these should be reported because they are expected to be overridden
- `private`: these should be reported because we can't override them (matches current behavior)
- `fileprivate`, `internal`, `public`: for this we would need to know if a function is being overridden within the same file/module. I don't know if SwiftLint already has this information when processing the AST, but usually linters work on individual files, so there is probably not much we can figure out about `internal` and `public` functions. Overriding for `fileprivate` might be determined by only looking at the file, but I'm not sure if it's worth the overhead.

**Comment by SimplyDanny** [comment]
I think that the assumption is that
```swift
class Base {
    func greet() { fatalError("Subclasses must override greet()") }
}
```
is unclean Swift. It mimics abstract or virtual classes but Swift doesn't have these concepts. The Swift-way would be the use of protocols.

I agree on `open` and `private` (but who would mark a private method as `@unavailable`?). SwiftLint is unable to check all others as you rightly describe. There are Analyzer rules which could support that. But they require more work to set them up and they are by magnitude slower than normal rules.

That said, I'd probably take your described case as a reason to restructure the code in a different way and not follow the rule's suggestion blindly.
